cloud/services/domains.go

```diff
@@ -0,0 +1,200 @@
+		if err != nil {
+			logger.Error(err, "Failed to update domain record")
+			return err
+		}
+	}
+
+	return nil
+}
+
+// DeleteNodeFromNB removes a backend Node from the Node Balancer configuration
+func DeleteIPFromDNS(
+	ctx context.Context,
+	logger logr.Logger,
+	machineScope *scope.MachineScope,
+) error {
+	// Check if instance is a control plane node
+	if !kutil.IsControlPlaneMachine(machineScope.Machine) {
+		return nil
+	}
+
+	// Get the public IP that was assigned
+	publicIP, err := GetMachinePublicIP(ctx, logger, machineScope)
+	if err != nil {
+		logger.Error(err, "Failed to get public IP of machine")
+		return err
+	}
+
+	// Get domainID from domain name
+	domainID, err := GetDomainID(ctx, logger, machineScope)
+	if err != nil {
+		logger.Error(err, "Failed to get domain ID")
+		return err
+	}
+
+	// Check if record exists for this IP and name combo
+	domainHostname := machineScope.LinodeCluster.ObjectMeta.Name + "-" + machineScope.LinodeCluster.Spec.Network.DNSUniqueIdentifier
+	filter, err := json.Marshal(map[string]interface{}{"name": domainHostname, "target": publicIP})
+	if err != nil {
+		logger.Error(err, "Failed to marshal domain filter")
+		return err
+	}
+
+	domainRecords, err := machineScope.LinodeDomainsClient.ListDomainRecords(ctx, domainID, linodego.NewListOptions(0, string(filter)))
+	if err != nil {
+		logger.Error(err, "unable to get current DNS record from API")
+		return err
+	}
+
+	// If domain record exists, delete it
+	if domainRecords != nil {
+		err := machineScope.LinodeDomainsClient.DeleteDomainRecord(ctx, domainID, domainRecords[0].ID)
+		if err != nil {
+			logger.Error(err, "Failed to delete domain record")
+			return err
+		}
+	}
+
+	if machineScope.LinodeCluster.Spec.Network.DNSTTLSec != 0 {
+		dnsTTLSec = machineScope.LinodeCluster.Spec.Network.DNSTTLSec
+	}
+
+	// Wait for TTL to expire
+	time.Sleep(time.Duration(dnsTTLSec) * time.Second)
+
+	return nil
+}
+
+// GetMachinePublicIP gets the machines public IP
+func GetMachinePublicIP(
+	ctx context.Context,
+	logger logr.Logger,
+	machineScope *scope.MachineScope,
+) (string, error) {
+	// Get the public IP that was assigned
+	addresses, err := machineScope.LinodeClient.GetInstanceIPAddresses(ctx, *machineScope.LinodeMachine.Spec.InstanceID)
+	if err != nil {
+		logger.Error(err, "Failed get instance IP addresses")
+
+		return "", err
+	}
+	if len(addresses.IPv4.Public) == 0 {
+		err := errors.New("no public IP address")
+		logger.Error(err, "no public IPV4 addresses set for LinodeInstance")
+
+		return "", err
+	}
+
+	return addresses.IPv4.Public[0].Address, nil
+}
+
+// GetDomainID gets the domains linode id
+func GetDomainID(
+	ctx context.Context,
+	logger logr.Logger,
+	machineScope *scope.MachineScope,
+) (int, error) {
+	// Get domainID from domain name
+	rootDomain := machineScope.LinodeCluster.Spec.Network.DNSRootDomain
+	filter, err := json.Marshal(map[string]interface{}{"domain": rootDomain})
```

Comment:
Please be explicit about types, when possible:

```suggestion
	filter, err := json.Marshal(map[string]string{"domain": rootDomain})
```

Reply:
updated